src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -193,48 +200,88 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
      * </ol>
      *
      * @param programmingExercise The programmingExercise that should be setup
-     * @param isImportedFromFile  defines if the programming exercise is imported from a file
+     * @param isImportedFromFile  defines if the programming exercise is imported from a file, if the
+     *                                exercise is imported, the build plans will not be triggered to prevent erroneous builds
      * @return The new setup exercise
      * @throws GitAPIException If something during the communication with the remote Git repository went wrong
      * @throws IOException     If the template files couldn't be read
      */
-    @Transactional // TODO: apply the transaction on a smaller scope
-    // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
     public ProgrammingExercise createProgrammingExercise(ProgrammingExercise programmingExercise, boolean isImportedFromFile) throws GitAPIException, IOException {
-        programmingExercise.generateAndSetProjectKey();
         final User exerciseCreator = userRepository.getUser();
-
         VersionControlService versionControl = versionControlService.orElseThrow();
-        programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
-        programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
-        initParticipations(programmingExercise);
-        setURLsAndBuildPlanIDsForNewExercise(programmingExercise);
-
-        // Save participations to get the ids required for the webhooks
-        connectBaseParticipationsToExerciseAndSave(programmingExercise);
-
-        connectAuxiliaryRepositoriesToExercise(programmingExercise);
 
-        programmingExerciseRepositoryService.setupExerciseTemplate(programmingExercise, exerciseCreator);
-        programmingSubmissionService.createInitialSubmissions(programmingExercise);
+        programmingExercise.generateAndSetProjectKey();
+        programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
+        ProgrammingExercise createdExercise;
+
+        try {
+            // We execute the following steps in a transaction to be able to roll back in the case of errors.
+            // As a transactional method cannot be used in the same class, we have to use this workaround.
+            createdExercise = transactionTemplate.execute(new TransactionCallback<ProgrammingExercise>() {
```

Comment:
Please let's avoid this pattern and try to use the annotation on a method.
Are you sure a transactional method cannot be used in the same class. I am not aware of this limitation

Reply:
alternatively, we could extract some parts (which need `@Transactional`) into another service

Reply:
I would also hope the required transactional context is even smaller, e.g. `programmingExerciseRepositoryService.createRepositoriesForNewExercise` should not require a DB transaction, does it?

Reply:
maybe it would be enough to put `connectBaseParticipationsToExerciseAndSave(programmingExercise);` together with an additional save of the programming exercise into a transactional context and then execute the rest without additional save operations of the programming exercise (maybe only once at the end?)